File: backend/scanner.py
import os
import sys
import glob
import re

# Conditionally import winreg for Windows registry access
try:
    import winreg
except ImportError:
    winreg = None

# Standard Native Linux plugin paths
LINUX_PLUGIN_PATHS = {
    "VST2": [
        os.path.expanduser("~/.vst"),
        "/usr/lib/vst",
        "/usr/local/lib/vst"
    ],
    "VST3": [
        os.path.expanduser("~/.vst3"),
        "/usr/lib/vst3",
        "/usr/local/lib/vst3"
    ],
    "CLAP": [
        os.path.expanduser("~/.clap"),
        "/usr/lib/clap",
        "/usr/local/lib/clap"
    ]
}

# Standard Native Windows plugin paths
WINDOWS_PLUGIN_PATHS = {
    "VST2": [
        "C:\\Program Files\\VSTPlugins",
        "C:\\Program Files\\Steinberg\\VSTPlugins",
        "C:\\Program Files (x86)\\VSTPlugins",
        "C:\\Program Files (x86)\\Steinberg\\VSTPlugins"
    ],
    "VST3": [
        "C:\\Program Files\\Common Files\\VST3",
        "C:\\Program Files (x86)\\Common Files\\VST3"
    ],
    "CLAP": [
        "C:\\Program Files\\Common Files\\CLAP",
        "C:\\Program Files (x86)\\Common Files\\CLAP"
    ],
    "AAX": [
        "C:\\Program Files\\Common Files\\Avid\\Audio\\Plug-Ins",
        "C:\\Program Files (x86)\\Common Files\\Avid\\Audio\\Plug-Ins"
    ]
}

# Standard relative paths within a Wine prefix (for Linux Wine scanner)
WINE_RELATIVE_PATHS = {
    "VST2": [
        "drive_c/Program Files/VSTPlugins",
        "drive_c/Program Files/Steinberg/VSTPlugins",
        "drive_c/Program Files (x86)/VSTPlugins"
    ],
    "WINE_VST3": [
        "drive_c/Program Files/Common Files/VST3",
        "drive_c/Program Files (x86)/Common Files/VST3"
    ],
    "WINE_CLAP": [
        "drive_c/Program Files/Common Files/CLAP",
        "drive_c/Program Files (x86)/Common Files/CLAP"
    ],
    "WINE_AAX": [
        "drive_c/Program Files/Common Files/Avid/Audio/Plug-Ins",
        "drive_c/Program Files (x86)/Common Files/Avid/Audio/Plug-Ins"
    ]
}

# Common Windows folders to look for presets/licenses
WINDOWS_RESOURCE_DIRS = [
    "C:\\ProgramData",
    os.path.expandvars("%USERPROFILE%\\AppData\\Roaming"),
    os.path.expandvars("%USERPROFILE%\\AppData\\Local"),
    os.path.expandvars("%USERPROFILE%\\Documents"),
    "C:\\Users\\Public\\Documents"
]

# Common folders within Wine prefix (Linux)
WINE_RESOURCE_DIRS = [
    "drive_c/ProgramData",
    "drive_c/users/Public/Documents",
    "drive_c/users/{user}/AppData/Roaming",
    "drive_c/users/{user}/AppData/Local",
    "drive_c/users/{user}/Documents"
]

LINUX_RESOURCE_DIRS = [
    os.path.expanduser("~/.config"),
    os.path.expanduser("~/.local/share"),
    os.path.expanduser("~/Documents")
]

def get_wine_prefixes():
    """Detect common Wine prefixes on the system (Linux only)."""
    if sys.platform == "win32":
        return [] # No wine prefixes on Windows native

    prefixes = []
    default_prefix = os.path.expanduser("~/.wine")
    if os.path.exists(default_prefix) and os.path.isdir(default_prefix):
        prefixes.append(default_prefix)
    
    bottles_path = os.path.expanduser("~/.var/app/com.usebottles.bottles/data/bottles/bottles")
    if os.path.exists(bottles_path):
        for bottle in os.listdir(bottles_path):
            bp = os.path.join(bottles_path, bottle)
            if os.path.isdir(bp) and os.path.exists(os.path.join(bp, "drive_c")):
                prefixes.append(bp)

    lutris_path = os.path.expanduser("~/Games")
    if os.path.exists(lutris_path):
        for game in os.listdir(lutris_path):
            gp = os.path.join(lutris_path, game)
            if os.path.isdir(gp) and os.path.exists(os.path.join(gp, "drive_c")):
                prefixes.append(gp)

    unique_prefixes = []
    for p in prefixes:
        abs_p = os.path.abspath(p)
        if abs_p not in unique_prefixes:
            unique_prefixes.append(abs_p)

    return unique_prefixes

def find_wine_users(prefix):
    """Find user directories in a Wine prefix."""
    users_path = os.path.join(prefix, "drive_c", "users")
    if not os.path.exists(users_path):
        return ["crossover", "steamuser", "default"]
    
    users = []
    for entry in os.listdir(users_path):
        full_path = os.path.join(users_path, entry)
        if os.path.isdir(full_path) and entry not in ["Public", "All Users", "Default User"]:
            users.append(entry)
    return users or ["crossover"]

def scan_native_plugins():
    """Scan native plugin directories for current platform."""
    plugins = []
    
    # Determine OS paths
    if sys.platform == "win32":
        paths_dict = WINDOWS_PLUGIN_PATHS
    else:
        paths_dict = LINUX_PLUGIN_PATHS
        
    for fmt, paths in paths_dict.items():
        for path in paths:
            if not os.path.exists(path):
                continue
            
            try:
                for entry in os.listdir(path):
                    full_path = os.path.join(path, entry)
                    is_plugin = False
                    
                    if fmt == "VST3" and (entry.lower().endswith(".vst3") or os.path.isdir(full_path)):
                        is_plugin = True
                    elif fmt == "CLAP" and entry.lower().endswith(".clap"):
                        is_plugin = True
                    elif fmt == "AAX" and (entry.lower().endswith(".aaxplugin") or os.path.isdir(full_path)):
                        is_plugin = True
                    elif fmt == "VST2" and (entry.lower().endswith(".so") or entry.lower().endswith(".dll")):
                        is_plugin = True
                    
                    if is_plugin:
                        plugins.append({
                            "name": os.path.splitext(entry)[0],
                            "format": fmt,
                            "path": full_path,
                            "type": "native",
                            "size": get_dir_size(full_path) if os.path.isdir(full_path) else os.path.getsize(full_path)
                        })
            except Exception:
                pass
    return plugins

def scan_wine_plugins(prefix):
    """Scan a specific Wine prefix for Windows plugins (Linux only)."""
    if sys.platform == "win32":
        return []
        
    plugins = []
    if not os.path.exists(prefix):
        return plugins
        
    users = find_wine_users(prefix)
    
    for fmt, rel_paths in WINE_RELATIVE_PATHS.items():
        for rel_path in rel_paths:
            full_search_path = os.path.join(prefix, rel_path)
            if not os.path.exists(full_search_path):
                continue
                
            try:
                for entry in os.listdir(full_search_path):
                    full_path = os.path.join(full_search_path, entry)
                    is_plugin = False
                    
                    if entry.endswith(".vst3") or entry.endswith(".dll") or entry.endswith(".clap") or entry.endswith(".aaxplugin") or os.path.isdir(full_path):
                        if fmt == "WINE_VST3" and (entry.endswith(".vst3") or os.path.isdir(full_path)):
                            is_plugin = True
                        elif fmt == "WINE_CLAP" and entry.endswith(".clap"):
                            is_plugin = True
                        elif fmt == "WINE_AAX" and (entry.endswith(".aaxplugin") or os.path.isdir(full_path)):
                            is_plugin = True
                        elif fmt == "VST2" and entry.endswith(".dll"):
                            is_plugin = True
                    
                    if is_plugin:
                        plugins.append({
                            "name": os.path.splitext(entry)[0],
                            "format": fmt.replace("WINE_", ""),
                            "path": full_path,
                            "type": "wine",
                            "prefix": prefix,
                            "size": get_dir_size(full_path) if os.path.isdir(full_path) else os.path.getsize(full_path)
                        })
            except Exception:
                pass
    return plugins

def get_dir_size(path):
    """Calculate total size of a directory in bytes."""
    total = 0
    try:
        if os.path.islink(path):
            return 0
        if os.path.isfile(path):
            return os.path.getsize(path)
        for dirpath, dirnames, filenames in os.walk(path):
            for f in filenames:
                fp = os.path.join(dirpath, f)
                if not os.path.islink(fp):
                    total += os.path.getsize(fp)
    except Exception:
        pass
    return total

def find_associated_resources(plugin_name, is_wine=False, prefix=None):
    """Scan common folders to suggest resource directories matching the plugin name."""
    suggestions = []
    name_clean = re.sub(r'[^a-zA-Z0-9]', '', plugin_name).lower()
    
    if sys.platform == "win32":
        # 1. Native Windows resources
        for r_dir in WINDOWS_RESOURCE_DIRS:
            if not os.path.exists(r_dir):
                continue
            try:
                for entry in os.listdir(r_dir):
                    entry_clean = re.sub(r'[^a-zA-Z0-9]', '', entry).lower()
                    if name_clean in entry_clean or entry_clean in name_clean:
                        suggestions.append({
                            "path": os.path.abspath(os.path.join(r_dir, entry)),
                            "description": f"Windows {os.path.basename(r_dir)} folder matching '{entry}'"
                        })
            except Exception:
                pass
                
        # 2. Windows registry suggestions
        if winreg:
            for hive_name, hive in [("HKEY_CURRENT_USER", winreg.HKEY_CURRENT_USER), ("HKEY_LOCAL_MACHINE", winreg.HKEY_LOCAL_MACHINE)]:
                try:
                    with winreg.OpenKey(hive, "Software") as key:
                        idx = 0
                        while True:
                            sub_name = winreg.EnumKey(key, idx)
                            sub_clean = re.sub(r'[^a-zA-Z0-9]', '', sub_name).lower()
                            if name_clean in sub_clean or sub_clean in name_clean:
                                suggestions.append({
                                    "path": f"{hive_name}\\Software\\{sub_name}",
                                    "description": f"Windows Registry key suggestion"
                                })
                            idx += 1
                except OSError:
                    pass # Reached end of registry subkeys
    else:
        # Linux / Wine
        if is_wine and prefix:
            users = find_wine_users(prefix)
            for user in users:
                for r_dir in WINE_RESOURCE_DIRS:
                    formatted_dir = r_dir.format(user=user)
                    full_base_path = os.path.join(prefix, formatted_dir)
                    if not os.path.exists(full_base_path):
                        continue
                    
                    try:
                        for entry in os.listdir(full_base_path):
                            entry_clean = re.sub(r'[^a-zA-Z0-9]', '', entry).lower()
                            if name_clean in entry_clean or entry_clean in name_clean:
                                suggestions.append({
                                    "path": os.path.abspath(os.path.join(full_base_path, entry)),
                                    "description": f"Wine {r_dir.split('/')[-1]} folder matching '{entry}'"
                                })
                    except Exception:
                        pass
        else:
            for r_dir in LINUX_RESOURCE_DIRS:
                if not os.path.exists(r_dir):
                    continue
                try:
                    for entry in os.listdir(r_dir):
                        entry_clean = re.sub(r'[^a-zA-Z0-9]', '', entry).lower()
                        if name_clean in entry_clean or entry_clean in name_clean:
                            suggestions.append({
                                "path": os.path.abspath(os.path.join(r_dir, entry)),
                                "description": f"Linux {r_dir.split('/')[-1]} folder matching '{entry}'"
                            })
                except Exception:
                    pass
                
    return suggestions

File: lib/services/scanner.dart
import 'dart:io';
import 'package:path/path.dart' as p;
import '../models/plugin.dart';

class ScannerService {
  // Standard Native Linux plugin paths
  static final Map<String, List<String>> linuxPaths = {
    'VST2': [
      '${Platform.environment['HOME']}/.vst',
      '/usr/lib/vst',
      '/usr/local/lib/vst',
    ],
    'VST3': [
      '${Platform.environment['HOME']}/.vst3',
      '/usr/lib/vst3',
      '/usr/local/lib/vst3',
    ],
    'CLAP': [
      '${Platform.environment['HOME']}/.clap',
      '/usr/lib/clap',
      '/usr/local/lib/clap',
    ],
  };

  // Standard Native Windows plugin paths
  static final Map<String, List<String>> windowsPaths = {
    'VST2': [
      'C:\\Program Files\\VSTPlugins',
      'C:\\Program Files\\Steinberg\\VSTPlugins',
      'C:\\Program Files (x86)\\VSTPlugins',
      'C:\\Program Files (x86)\\Steinberg\\VSTPlugins',
    ],
    'VST3': [
      'C:\\Program Files\\Common Files\\VST3',
      'C:\\Program Files (x86)\\Common Files\\VST3',
    ],
    'CLAP': [
      'C:\\Program Files\\Common Files\\CLAP',
      'C:\\Program Files (x86)\\Common Files\\CLAP',
    ],
    'AAX': [
      'C:\\Program Files\\Common Files\\Avid\\Audio\\Plug-Ins',
      'C:\\Program Files (x86)\\Common Files\\Avid\\Audio\\Plug-Ins',
    ],
  };

  // Standard Native macOS plugin paths
  static final Map<String, List<String>> macosPaths = {
    'VST2': [
      '/Library/Audio/Plug-Ins/VST',
      '${Platform.environment['HOME']}/Library/Audio/Plug-Ins/VST',
    ],
    'VST3': [
      '/Library/Audio/Plug-Ins/VST3',
      '${Platform.environment['HOME']}/Library/Audio/Plug-Ins/VST3',
    ],
    'CLAP': [
      '/Library/Audio/Plug-Ins/CLAP',
      '${Platform.environment['HOME']}/Library/Audio/Plug-Ins/CLAP',
    ],
    'AAX': [
      '/Library/Application Support/Avid/Audio/Plug-Ins',
    ],
  };

  // Standard relative paths within a Wine prefix (for Linux Wine scanner)
  static final Map<String, List<String>> wineRelativePaths = {
    'VST2': [
      'drive_c/Program Files/VSTPlugins',
      'drive_c/Program Files/Steinberg/VSTPlugins',
      'drive_c/Program Files (x86)/VSTPlugins',
    ],
    'WINE_VST3': [
      'drive_c/Program Files/Common Files/VST3',
      'drive_c/Program Files (x86)/Common Files/VST3',
    ],
    'WINE_CLAP': [
      'drive_c/Program Files/Common Files/CLAP',
      'drive_c/Program Files (x86)/Common Files/CLAP',
    ],
    'WINE_AAX': [
      'drive_c/Program Files/Common Files/Avid/Audio/Plug-Ins',
      'drive_c/Program Files (x86)/Common Files/Avid/Audio/Plug-Ins',
    ],
  };

  // Common Windows resource folders to look for presets/licenses
  static final List<String> windowsResourceDirs = [
    'C:\\ProgramData',
    'C:\\Users\\Public\\Documents',
    'APPDATA\\Roaming', // Resolved dynamically relative to UserProfile
    'APPDATA\\Local',
    'DOCUMENTS',
  ];

  // Common folders within Wine prefix (Linux)
  static final List<String> wineResourceDirs = [
    'drive_c/ProgramData',
    'drive_c/users/Public/Documents',
    'drive_c/users/{user}/AppData/Roaming',
    'drive_c/users/{user}/AppData/Local',
    'drive_c/users/{user}/Documents',
  ];

  static final List<String> linuxResourceDirs = [
    '${Platform.environment['HOME']}/.config',
    '${Platform.environment['HOME']}/.local/share',
    '${Platform.environment['HOME']}/Documents',
  ];

  // Detect common Wine prefixes on the system (Linux only)
  List<String> getWinePrefixes() {
    if (!Platform.isLinux) return [];

    final List<String> prefixes = [];
    final home = Platform.environment['HOME'] ?? '';
    if (home.isEmpty) return [];

    // 1. Default Wine prefix
    final defaultPrefix = p.join(home, '.wine');
    if (Directory(defaultPrefix).existsSync()) {
      prefixes.add(defaultPrefix);
    }

    // 2. Bottles
    final bottlesPath = p.join(home, '.var', 'app', 'com.usebottles.bottles', 'data', 'bottles', 'bottles');
    final bottlesDir = Directory(bottlesPath);
    if (bottlesDir.existsSync()) {
      try {
        for (var entity in bottlesDir.listSync()) {
          if (entity is Directory && Directory(p.join(entity.path, 'drive_c')).existsSync()) {
            prefixes.add(entity.path);
          }
        }
      } catch (_) {}
    }

    // 3. Lutris
    final lutrisPath = p.join(home, 'Games');
    final lutrisDir = Directory(lutrisPath);
    if (lutrisDir.existsSync()) {
      try {
        for (var entity in lutrisDir.listSync()) {
          if (entity is Directory && Directory(p.join(entity.path, 'drive_c')).existsSync()) {
            prefixes.add(entity.path);
          }
        }
      } catch (_) {}
    }

    return prefixes.toSet().toList(); // Unique resolved paths
  }

  List<String> findWineUsers(String prefix) {
    final usersPath = p.join(prefix, 'drive_c', 'users');
    final usersDir = Directory(usersPath);
    if (!usersDir.existsSync()) {
      return ['crossover', 'steamuser', 'default'];
    }

    final List<String> users = [];
    try {
      for (var entry in usersDir.listSync()) {
        final name = p.basename(entry.path);
        if (entry is Directory && name != 'Public' && name != 'All Users' && name != 'Default User') {
          users.add(name);
        }
      }
    } catch (_) {}
    return users.isEmpty ? ['crossover'] : users;
  }

  // Scan local plugins installed natively on this system
  List<PluginModel> scanNativePlugins() {
    final List<PluginModel> plugins = [];
    Map<String, List<String>> pathsDict;

    if (Platform.isWindows) {
      pathsDict = windowsPaths;
    } else if (Platform.isMacOS) {
      pathsDict = macosPaths;
    } else if (Platform.isLinux) {
      pathsDict = linuxPaths;
    } else {
      return []; // Unsupported (e.g. iOS standalone scanning not allowed)
    }

    pathsDict.forEach((fmt, paths) {
      for (var path in paths) {
        final dir = Directory(path);
        if (!dir.existsSync()) continue;

        try {
          for (var entry in dir.listSync()) {
            final entryName = p.basename(entry.path);
            final nameLower = entryName.toLowerCase();
            bool isPlugin = false;

            if (fmt == 'VST3' && (nameLower.endsWith('.vst3') || entry is Directory)) {
              isPlugin = true;
            } else if (fmt == 'CLAP' && nameLower.endsWith('.clap')) {
              isPlugin = true;
            } else if (fmt == 'AAX' && (nameLower.endsWith('.aaxplugin') || entry is Directory)) {
              isPlugin = true;
            } else if (fmt == 'VST2' && (nameLower.endsWith('.so') || nameLower.endsWith('.dll') || nameLower.endsWith('.vst'))) {
              isPlugin = true;
            }

            if (isPlugin) {
              int size = 0;
              try {
                size = getDirSize(entry.path);
              } catch (_) {}

              plugins.add(PluginModel(
                name: p.basenameWithoutExtension(entry.path),
                format: fmt,
                path: entry.path,
                type: 'native',
                size: size,
              ));
            }
          }
        } catch (_) {}
      }
    });

    return plugins;
  }

  // Scan plugins inside a specific Wine prefix (Linux/macOS)
  List<PluginModel> scanWinePlugins(String prefix) {
    if (Platform.isWindows) return [];

    final List<PluginModel> plugins = [];
    final prefixDir = Directory(prefix);
    if (!prefixDir.existsSync()) return [];

    wineRelativePaths.forEach((fmt, relPaths) {
      for (var relPath in relPaths) {
        final fullPath = p.join(prefix, relPath);
        final dir = Directory(fullPath);
        if (!dir.existsSync()) continue;

        try {
          for (var entry in dir.listSync()) {
            final entryName = p.basename(entry.path);
            bool isPlugin = false;

            if (entryName.endsWith('.vst3') || entryName.endsWith('.dll') || entryName.endsWith('.clap') || entryName.endsWith('.aaxplugin') || entry is Directory) {
              if (fmt == 'WINE_VST3' && (entryName.endsWith('.vst3') || entry is Directory)) {
                isPlugin = true;
              } else if (fmt == 'WINE_CLAP' && entryName.endsWith('.clap')) {
                isPlugin = true;
              } else if (fmt == 'WINE_AAX' && (entryName.endsWith('.aaxplugin') || entry is Directory)) {
                isPlugin = true;
              } else if (fmt == 'VST2' && entryName.endsWith('.dll')) {
                isPlugin = true;
              }
            }

            if (isPlugin) {
              int size = 0;
              try {
                size = getDirSize(entry.path);
              } catch (_) {}

              plugins.add(PluginModel(
                name: p.basenameWithoutExtension(entry.path),
                format: fmt.replaceFirst('WINE_', ''),
                path: entry.path,
                type: 'wine',
                prefix: prefix,
                size: size,
              ));
            }
          }
        } catch (_) {}
      }
    });

    return plugins;
  }

  int getDirSize(String path) {
    final file = File(path);
    if (file.existsSync()) return file.lengthSync();

    final dir = Directory(path);
    if (!dir.existsSync()) return 0;

    int totalSize = 0;
    try {
      for (var entity in dir.listSync(recursive: true)) {
        if (entity is File) {
          totalSize += entity.lengthSync();
        }
      }
    } catch (_) {}
    return totalSize;
  }

  // Find associated folders & registry keys matching a plugin name
  Future<List<Map<String, String>>> findAssociatedResources(String pluginName, {bool isWine = false, String? prefix}) async {
    final List<Map<String, String>> suggestions = [];
    final nameClean = pluginName.replaceAll(RegExp(r'[^a-zA-Z0-9]'), '').toLowerCase();

    if (Platform.isWindows) {
      final userProfile = Platform.environment['USERPROFILE'] ?? '';

      // 1. Traverse directories
      for (var rDir in windowsResourceDirs) {
        String resolvedDir = rDir;
        if (rDir.startsWith('APPDATA\\Roaming')) {
          resolvedDir = p.join(userProfile, 'AppData', 'Roaming');
        } else if (rDir.startsWith('APPDATA\\Local')) {
          resolvedDir = p.join(userProfile, 'AppData', 'Local');
        } else if (rDir.startsWith('DOCUMENTS')) {
          resolvedDir = p.join(userProfile, 'Documents');
        }

        final dir = Directory(resolvedDir);
        if (!dir.existsSync()) continue;

        try {
          for (var entry in dir.listSync()) {
            final entryName = p.basename(entry.path);
            final entryClean = entryName.replaceAll(RegExp(r'[^a-zA-Z0-9]'), '').toLowerCase();
            if (entryClean.contains(nameClean) || nameClean.contains(entryClean)) {
              suggestions.add({
                'path': entry.path,
                'description': 'Windows ${p.basename(resolvedDir)} folder matching "$entryName"',
              });
            }
          }
        } catch (_) {}
      }

      // 2. Scan Registry via Command Line reg query (UAC safe!)
      for (var hive in ['HKCU\\Software', 'HKLM\\Software']) {
        try {
          var result = await Process.run('reg', ['query', hive]);
          if (result.exitCode == 0) {
            var output = result.stdout.toString();
            for (var line in output.split('\n')) {
              var trimmed = line.trim();
              if (trimmed.isEmpty) continue;
              var keyName = p.basename(trimmed);
              var keyClean = keyName.replaceAll(RegExp(r'[^a-zA-Z0-9]'), '').toLowerCase();
              if (keyClean.contains(nameClean) || nameClean.contains(keyClean)) {
                suggestions.add({
                  'path': trimmed,
                  'description': 'Windows Registry Key Suggestion',
                });
              }
            }
          }
        } catch (_) {}
      }
    } else {
      // Linux or macOS
      if (isWine && prefix != null) {
        final users = findWineUsers(prefix);
        for (var user in users) {
          for (var rDir in wineResourceDirs) {
            final formattedDir = rDir.replaceFirst('{user}', user);
            final fullPath = p.join(prefix, formattedDir);
            final dir = Directory(fullPath);
            if (!dir.existsSync()) continue;

            try {
              for (var entry in dir.listSync()) {
                final entryName = p.basename(entry.path);
                final entryClean = entryName.replaceAll(RegExp(r'[^a-zA-Z0-9]'), '').toLowerCase();
                if (entryClean.contains(nameClean) || nameClean.contains(entryClean)) {
                  suggestions.add({
                    'path': entry.path,
                    'description': 'Wine ${rDir.split('/').last} folder matching "$entryName"',
                  });
                }
              }
            } catch (_) {}
          }
        }
      } else {
        // Native Linux/macOS resource suggestions
        for (var rDir in linuxResourceDirs) {
          final dir = Directory(rDir);
          if (!dir.existsSync()) continue;

          try {
            for (var entry in dir.listSync()) {
              final entryName = p.basename(entry.path);
              final entryClean = entryName.replaceAll(RegExp(r'[^a-zA-Z0-9]'), '').toLowerCase();
              if (entryClean.contains(nameClean) || nameClean.contains(entryClean)) {
                suggestions.add({
                  'path': entry.path,
                  'description': 'Local ${p.basename(rDir)} folder matching "$entryName"',
                });
              }
            }
          } catch (_) {}
        }
      }
    }

    return suggestions;
  }
}
